Define delayed event ratelimit category

AndrewFerr opened the pull request:
Apply ratelimiting on delayed event management separately from messages.

MSC4140

### Pull Request Checklist

<!-- Please read https://element-hq.github.io/synapse/latest/development/contributing_guide.html before submitting your pull request -->

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [ ] [Code style](https://element-hq.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

AndrewFerr committed:
Define delayed event ratelimit category

Apply ratelimiting on delayed event management separately from messages.

AndrewFerr committed:
Add changelog

hughns commented on docs/usage/configuration/config_documentation.md:
```diff
@@ -1866,6 +1866,25 @@ rc_federation:
+This is a ratelimiting option that ratelimits
+attempts to restart, cancel, or view delayed events
+based on the account the client is using.
+It defaults to: `per_second: 10`, `burst_count: 100`.
```
Why are the defaults so different from `rc_message` which is what is currently used? I expect the default for `rc_delayed_event` to be similar.

AndrewFerr replied:
The default is more lax to permit refreshing delayed events often / in a loop without risk of ratelimiting, whereas "actual" messages are sent less frequently / not periodically and benefit from a stricter ratelimit.

AndrewFerr replied:
b2b1e228fd

hughns reviewed (commented):
Please see question inline about the default value.

anoadragon453 commented on tests/rest/client/test_delayed_events.py:
```diff
@@ -206,6 +225,44 @@ def test_cancel_delayed_state_event(self) -> None:
+    @unittest.override_config({"rc_delayed_event": {"per_second": 0.5, "burst_count": 1}})
+    def test_cancel_delayed_event_ratelimit(self) -> None:
+        delay_ids = []
+        for i in range(2):
```
```suggestion
        for _ in range(2):
```

AndrewFerr replied:
b1dd548401

fkwp commented:
@AndrewFerr what is the current status on this topic?

AndrewFerr commented:
> @AndrewFerr what is the current status on this topic?

I still need to address the review comments, but once that's done, this should be good to go (barring further review).  It should also be usable for testing/development until then.

Since this ratelimit targets the simpler management commands of cancel/restart/get, it's less complicated than [the ratelimits we want for delayed messages themselves](https://github.com/element-hq/synapse/pull/18018), which does require more thought.

AndrewFerr committed:
Link comment to GH issue it relates to

Co-authored-by: Andrew Morgan <1342360+anoadragon453@users.noreply.github.com>

AndrewFerr committed:
Use _ for unreferenced loop variables in test

AndrewFerr commented on synapse/config/ratelimiting.py:
```diff
@@ -228,3 +228,9 @@ def read_config(self, config: JsonDict, **kwargs: Any) -> None:
+        self.rc_delayed_event = RatelimitSettings.parse(
+            config,
+            "rc_delayed_event",
+            defaults={"per_second": 10, "burst_count": 100},
```
Fair enough.  For call connection management, the js-sdk defaults to sending a refresh once every 8 seconds, so the high ratelimit was just to be sure that it would never get in the way.

I like the idea of a per-device limit rather than a high allowed rate.  That permits a usecase of many devices of the same user connected to the same call, as they'd each need to send their own call refreshes & would get blocked if the ratelimit were per-user.

AndrewFerr replied:
5b2ae48804

anoadragon453 replied:
Hrm, I'm against doing it on a per-device basis. If we do so, a user can register 100 devices and then suddenly send 100 events/s.

What's the use case for a single user being in the same call >3 times?

anoadragon453 replied:
Ah sorry, this is for event management, not for sending events.

In that case this is probably fine - it's not as heavy as actually sending events.

AndrewFerr replied:
I've renamed the config key to be more clear that it's just for management and not the events themselves.

AndrewFerr committed:
Document the purpose of a high rc_delayed_event

AndrewFerr committed:
Move creation of delayed event ratelimiter

AndrewFerr committed:
Key delayed event ratelimit per device ID

Also change the default values appropriately

AndrewFerr committed:
Add missing import

AndrewFerr committed:
Lint modified files

AndrewFerr committed:
Fix Ratelimiter construction

AndrewFerr committed:
Fix test config

AndrewFerr committed:
Merge branch 'develop' into msc4140_ratelimit-category

anoadragon453 commented on demo/start.sh:
```diff
@@ -138,6 +138,9 @@ for port in 8080 8081 8082; do
 			  per_user:
 			    per_second: 1000
 			    burst_count: 1000
+			rc_delayed_event:
+			  per_second: 1000
+			  burst_count: 1000
```
Note that this should also be added to Sytest's config in a PR on that repo. See this example: https://github.com/matrix-org/sytest/pull/1394

AndrewFerr replied:
https://github.com/matrix-org/sytest/pull/1395

anoadragon453 commented on synapse/handlers/delayed_events.py:
```diff
@@ -227,6 +237,9 @@ async def add(
             SynapseError: if the delayed event fails validation checks.
         """
+        # Use standard request limiter for scheduling new delayed events.
+        # TODO: Instead apply rateliming based on the scheduled send time.
```
```suggestion
        # TODO: Instead apply ratelimiting based on the scheduled send time.
```

anoadragon453 reviewed (approved):
This LGTM!

Complement test failures are a flake. Note that `TestDelayedEvents/delayed_state_events_are_kept_on_server_restart` has been flakey for a while; could you please fix that in another PR @AndrewFerr :)

anoadragon453 committed:
fix typo in comment

AndrewFerr committed:
Merge branch 'develop' into msc4140_ratelimit-category

AndrewFerr committed:
Rename ratelimit config key to clarify its purpose

This rate limit is for the _management_ of delayed events, not the
sending of the events themselves (which will be managed by a different
ratelimit setting to be added later).

AndrewFerr committed:
Merge branch 'develop' into msc4140_ratelimit-category

AndrewFerr committed:
Merge branch 'develop' into msc4140_ratelimit-category

AndrewFerr commented:
This should be ready to merge now.  The only change since the review is b014c078cb.
